cypress/integration/slots.spec.js

```diff
@@ -0,0 +1,79 @@
+/// <reference types="cypress" />
+
+import { getTestUrl } from '../lib/url';
+import { getIframeBody } from '../lib/iframe';
+
+// Don't fail tests when uncaught exceptions occur
+// This is because scripts loaded on the page and unrelated to these tests can cause this
+Cypress.on('uncaught:exception', (err, runnable) => {
+	return false;
+});
+
+// Pass different stage in via environment variable
+// e.g. `yarn cypress run --env stage=code`
+const stage = Cypress.env('stage');
+
+const pages = [
+	{
+		path: getTestUrl(
+			stage,
+			'/environment/2020/oct/13/maverick-rewilders-endangered-species-extinction-conservation-uk-wildlife',
+			{ isDcr: true },
+		),
+		expectedMinTotalSlotsOnPage: 14,
+		expectedMinInlineSlotsOnPage: 10,
+		adTest: 'fixed-puppies',
+	},
+];
+
+describe('Slots and iframes load on pages', () => {
+	pages.forEach(
+		({
+			path,
+			adTest,
+			expectedMinTotalSlotsOnPage,
+			expectedMinInlineSlotsOnPage,
+		}) => {
+			it(`Test ${path} has at least ${expectedMinInlineSlotsOnPage} inline and ${expectedMinTotalSlotsOnPage} total slots`, () => {
+				cy.visit(`${path}?adtest=${adTest}`);
+
+				// Click "Yes, I'm happy" on the sourcepoint banner to obtain consent
+				getIframeBody('sp_message_iframe_597005')
+					.find('.btn-primary', { timeout: 10_000 })
+					.click();
```

Comment:
Since the consent-management-platform has pretty reliable UI tests itself, I wonder if it would be worth mocking user consent somehow in order to reduce the duplicated logic and make this test more focused on ad slots.
I guess the down side is that the test would be a bit less realistic.

Reply:
I can understand the argument to keep the tests focussed on our code.

I personally would prefer to test consent (UI & functionality) as a part of our tests as its such an intertwined piece of functionality. Integration bugs could be missed downstream in Sourcepoint or CMP.

In general for e2e tests, I find the closer to realistic user behaviour the better they are at capturing real world bugs.

Reply:
I can see the benefits of both approaches. For example, if you're running an integration test against a payment flow you'll probably want to mock a user logging in rather than testing that flow as well. In that regard I can see why we might want to mock this consent stage. 

However, looking at the CMP tests, these only run against that repo? I think the advantage here of knowing at the point we run these tests that the consent banner is functioning as expected could be important.

Reply:
Both good points - I wasn't considering that this test is a bit broader than the CMP's UI tests which don't cover the package's use in frontend or anywhere else.